R/testthat-reporter.R

```diff
@@ -0,0 +1,147 @@
+
+globalVariables("private")
+
+#' testthat reporter that checks if child proceses are cleaned up in tests
+#'
+#' `CleanupReporter` takes an existing testthat `Reporter` object, and
+#' wraps it, so it checks for leftover child processes, at the specified
+#' place, see the `unit` argument below.
+#'
+#' Child processes can be reported via a failed expectation, cleaned up
+#' silently, or cleaned up and reptorted (the default).
+#'
+#' The constructor of the `CleanupReporter` class has options:
+#' * `file`: the output file, if any, this is passed to `reporter`.
+#' * `unit`: when to perform the child process check and cleanup. Possible
+#'   values:
+#'     * `"test"`: at the end of each [testthat::test_that()] block
+#'       (the default),
+#'     * `"testsuite"`: at the end of the test suite.
+#' * `process_cleanup`: Logical scalar, whether to kill the leftover
+#'   processes.
+#' * `process_fail`: Whether to create an expectation, that fails if there
+#'   are any processes alive.
+#' * `process_timeout`: How long to wait for the processes to quit. This is
+#'   sometimes needed, because even if some kill signals were sent to
+#'   child processes, it might take a short time for these to take effect.
+#'   It defaults to one second.
+#'
+#' @note Some IDEs, like RStudio, start child processes frequently, and
```

Comment:
Is there anything we can do about this? Are you basing this off of the pid of the whole RStudio application, or the pid of the embedded R terminal?

Reply:
The pid of the RStudio application is irrelevant, we don't deal with that. But the embedded R process (`rsession`) is also multi-theaded, and the other threads spawn processes quite frequently. This is by design, and will stay like this, as I understand. 

I don't know if the crash of the whole session, when its child processes are killed, should be considered a bug, and I don't know how hard it would be to fix this, but I'll report an issue.

Reply:
Btw. in RStudio, the shell of the terminal is also a child process of `rsession`, that's why terminals crash/reset whenever `rsession` crashes. So even if `rsession` will survive our  cleanup, at least the terminals will be gone....